client/state/data-layer/wpcom/sites/simple-payments/index.js

```diff
@@ -117,143 +140,69 @@ export function productToCustomPost( product ) {
-	const { siteId, product, productId } = action;
-
-	dispatch(
-		http(
-			{
-				method: 'POST',
-				path: `/sites/${ siteId }/posts/${ productId }`,
-				body: productToCustomPost( product ),
-			},
-			action
-		)
-	);
-}
-
-/**
- * Issues an API request to delete a product
- * @param {Object} store Redux store
- * @param {Object} action Action object
- */
-export function requestSimplePaymentsProductDelete( { dispatch }, action ) {
-	const { siteId, productId } = action;
-
-	dispatch(
-		http(
-			{
-				method: 'POST',
-				path: `/sites/${ siteId }/posts/${ productId }/delete`,
-			},
-			action
-		)
-	);
-}
-
-export const addProduct = ( { dispatch }, { siteId }, newProduct ) =>
-	dispatch( receiveUpdateProduct( siteId, customPostToProduct( newProduct ) ) );
-
-export const deleteProduct = ( { dispatch }, { siteId }, deletedProduct ) =>
-	dispatch( receiveDeleteProduct( siteId, deletedProduct.ID ) );
-
-export const listProduct = ( { dispatch }, { siteId }, product ) => {
-	if ( ! isValidSimplePaymentsProduct( product ) ) {
-		return;
-	}
-
-	dispatch( receiveProduct( siteId, customPostToProduct( product ) ) );
-};
-
-export const listProducts = ( { dispatch }, { siteId }, { posts: products } ) => {
-	const validProducts = filter( products, isValidSimplePaymentsProduct );
-
-	dispatch( receiveProductsList( siteId, validProducts.map( customPostToProduct ) ) );
-};
+		} ),
+	fromApi: customPostsToProducts,
+	onSuccess: replaceProductList,
+} );
+
+export const handleProductListAdd = dispatchRequest( {
+	initiator: action =>
+		http( {
+			method: 'POST',
+			path: `/sites/${ action.siteId }/posts/new`,
+			body: productToCustomPost( action.product ),
+		} ),
+	fromApi: customPostToProduct,
+	onSuccess: addOrUpdateProduct,
+} );
+
+export const handleProductListEdit = dispatchRequest( {
+	initiator: action =>
+		http( {
+			method: 'POST',
+			path: `/sites/${ action.siteId }/posts/${ action.productId }`,
+			body: productToCustomPost( action.product ),
+		} ),
+	fromApi: customPostToProduct,
+	onSuccess: addOrUpdateProduct,
+} );
+
+export const handleProductListDelete = dispatchRequest( {
+	initiator: action =>
+		http( {
+			method: 'POST',
+			path: `/sites/${ action.siteId }/posts/${ action.productId }/delete`,
+		} ),
+	onSuccess: deleteProduct,
+} );
 
 export default {
-	[ SIMPLE_PAYMENTS_PRODUCT_GET ]: [
-		dispatchRequest( requestSimplePaymentsProduct, listProduct, noop ),
-	],
-	[ SIMPLE_PAYMENTS_PRODUCTS_LIST ]: [
-		dispatchRequest( requestSimplePaymentsProducts, listProducts, noop ),
-	],
-	[ SIMPLE_PAYMENTS_PRODUCTS_LIST_ADD ]: [
-		dispatchRequest( requestSimplePaymentsProductAdd, addProduct, noop ),
-	],
-	[ SIMPLE_PAYMENTS_PRODUCTS_LIST_EDIT ]: [
-		dispatchRequest( requestSimplePaymentsProductEdit, addProduct, noop ),
-	],
-	[ SIMPLE_PAYMENTS_PRODUCTS_LIST_DELETE ]: [
-		dispatchRequest( requestSimplePaymentsProductDelete, deleteProduct, noop ),
-	],
+	[ SIMPLE_PAYMENTS_PRODUCT_GET ]: handleProductGet,
```

Comment:
need to wrap all the values in square brackets now that dispatchRequest doesn't return an array. i.e.:

```js
export default {
  [ SIMPLE_PAYMENTS_PRODUCT_GET ]: [ handleProductGet ]  
}
```

Reply:
Good catch! Fixed and tested 👍